src/Libraries/RevitServices/Events/EventHandlerProxy.cs

```diff
@@ -0,0 +1,168 @@
+            HandleEvent(EventType.DocumentOpened, "OnApplicationDocumentOpened", sender, args);
+        }
+
+        /// <summary>
+        /// The event handler of the proxy to be subscribed to the DocumentClosing event publisher
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="args"></param>
+        public virtual void OnApplicationDocumentClosing(object sender, DocumentClosingEventArgs args)
+        {
+            HandleEvent(EventType.DocumentClosing, "OnApplicationDocumentClosing", sender, args);
+        }
+
+        /// <summary>
+        /// The event handler of the proxy to be subscribed to the DocumentClosed event publisher
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="args"></param>
+        public virtual void OnApplicationDocumentClosed(object sender, DocumentClosedEventArgs args)
+        {
+            HandleEvent(EventType.DocumentClosed, "OnApplicationDocumentClosed", sender, args);
+        }
+
+        /// <summary>
+        /// The event handler of the proxy to be subscribed to the ViewActivating event publisher
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="args"></param>
+        public virtual void OnApplicationViewActivating(object sender, ViewActivatingEventArgs args)
+        {
+            HandleEvent(EventType.ViewActivating, "OnApplicationViewActivating", sender, args);
+        }
+
+        /// <summary>
+        /// The event handler of the proxy to be subscribed to the ViewActivated event publisher
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="args"></param>
+        public virtual void OnApplicationViewActivated(object sender, ViewActivatedEventArgs args)
+        {
+            HandleEvent(EventType.ViewActivated, "OnApplicationViewActivated", sender, args);
+        }
+
+        /// <summary>
+        /// This function will first search the dictionary to see whether a method for a given type of
+        /// event is in the dictionary or not. If so, it will invoke the method. Otherwise, it will try
+        /// to see whether a method of the given type with a specified name is there or not. If so, it will
+        /// invoke this method. If not, it will do nothing.
+        /// </summary>
+        /// <param name="type"></param>
+        /// <param name="defaultMethodName"></param>
+        /// <param name="sender"></param>
+        /// <param name="args"></param>
+        protected void HandleEvent(EventType type, string defaultMethodName, object sender, EventArgs args)
+        {
+            var h = _ref.Target;
+            if (h != null)
+            {
+                MethodInfo m;
+                if (_dict.TryGetValue(type, out m) && m != null)
+                {
+                    m.Invoke(h, new[] { sender, args });
+                }
+                else
+                {
+                    var t = h.GetType();
```

Comment:
For some cases, the proxy will only be used for 1 or 2 types of events, then I will add the delegate's methodinfo to the dictionary. For RevitDynamoModel, all 6 cases are used, so the methodinfo will be found through reflection.